plugins/obs-webrtc/whep-source.cpp

```diff
@@ -0,0 +1,511 @@
+
+	auto naluType = uint8_t(msg.at(0)) & naluTypeBitmask;
+	msg.insert(msg.begin(),
+		   {std::byte(0), std::byte(0), std::byte(0), std::byte(1)});
+
+	if (naluType == naluTypeSPS || naluType == naluTypePPS) {
+		std::move(msg.begin(), msg.end(),
+			  std::back_inserter(sps_and_pps));
+		return;
+	} else if (this->sps_and_pps.size() != 0) {
+		std::copy(this->sps_and_pps.begin(), this->sps_and_pps.end(),
+			  std::back_inserter(msg));
+		this->sps_and_pps = std::vector<std::byte>{};
+	}
+
+	pkt->data = reinterpret_cast<uint8_t *>(msg.data());
+	pkt->size = static_cast<int>(msg.size());
+
+	auto ret = avcodec_send_packet(this->video_av_codec_context.get(), pkt);
+	if (ret < 0) {
+		return;
+	}
+
+	AVFrame *av_frame = this->av_frame.get();
+
+	while (true) {
+		ret = avcodec_receive_frame(this->video_av_codec_context.get(),
+					    av_frame);
+		if (ret < 0) {
+			if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) {
+				break;
+			}
+			return;
+		}
+
+		struct obs_source_frame frame = {};
+
+		frame.format = VIDEO_FORMAT_I420;
+		frame.width = av_frame->width;
+		frame.height = av_frame->height;
+		frame.timestamp = pts;
+		frame.max_luminance = 0;
+		frame.trc = VIDEO_TRC_DEFAULT;
+
+		video_format_get_parameters_for_format(
+			VIDEO_CS_DEFAULT, VIDEO_RANGE_DEFAULT, frame.format,
+			frame.color_matrix, frame.color_range_min,
+			frame.color_range_max);
+
+		for (size_t i = 0; i < MAX_AV_PLANES; i++) {
+			frame.data[i] = av_frame->data[i];
+			frame.linesize[i] = abs(av_frame->linesize[i]);
+		}
+
+		obs_source_output_video(this->source, &frame);
+		last_frame = std::chrono::system_clock::now();
+	}
+}
+
+void WHEPSource::SetupPeerConnection()
+{
+	peer_connection = std::make_shared<rtc::PeerConnection>();
+	peer_connection->onStateChange([this](rtc::PeerConnection::State state) {
+		switch (state) {
+		case rtc::PeerConnection::State::New:
+			do_log(LOG_INFO, "PeerConnection state is now: New");
+			break;
+		case rtc::PeerConnection::State::Connecting:
+			do_log(LOG_INFO,
+			       "PeerConnection state is now: Connecting");
+			break;
+		case rtc::PeerConnection::State::Connected:
+			do_log(LOG_INFO,
+			       "PeerConnection state is now: Connected");
+			break;
+		case rtc::PeerConnection::State::Disconnected:
+			do_log(LOG_INFO,
+			       "PeerConnection state is now: Disconnected");
+			break;
+		case rtc::PeerConnection::State::Failed:
+			do_log(LOG_INFO, "PeerConnection state is now: Failed");
+			break;
+		case rtc::PeerConnection::State::Closed:
+			do_log(LOG_INFO, "PeerConnection state is now: Closed");
+			break;
+		}
+	});
+
+	rtc::Description::Audio audioMedia(
+		"0", rtc::Description::Direction::RecvOnly);
+	audioMedia.addOpusCodec(111);
+	audio_track = peer_connection->addTrack(audioMedia);
+
+	auto audio_depacketizer = std::make_shared<rtc::OpusRtpDepacketizer>();
+	auto audio_session = std::make_shared<rtc::RtcpReceivingSession>();
+	audio_session->addToChain(audio_depacketizer);
```

Comment:
I have discovered that the current WHEP client is not sending any RTCP packets (including PLI) to SFU at all because (maybe) `audio_session` is not passed to `audio_track`.

The issue has been resolved by the following change (`video_session` has the same issue though):
```suggestion
        audio_depacketizer->addToChain(audio_session);
```

Reply:
Fixed! Thank you for the review @sile

Reply:
@Sean-Der Let me notice that the latest revision still has the same problem.